### ISSUE TITLE
Softmax and Log Softmax and not supported in SMPC

### ISSUE THREAD

**Issue by LaRiffle** [OpenMined/PySyft#3232]
Infos about this functions:

⚠️ Beware of numerical stabililty when using exponentials! Suggested reading: https://timvieira.github.io/blog/post/2014/02/11/exp-normalize-trick/

### Softmax
![Capture d’écran 2020-03-21 à 15 01 44](https://user-images.githubusercontent.com/12446521/77228063-f8aa8000-6b84-11ea-813a-a2e484bf0697.png)
https://pytorch.org/docs/stable/nn.html#softmax

### Log Softmax
![Capture d’écran 2020-03-21 à 15 02 03](https://user-images.githubusercontent.com/12446521/77228061-f811e980-6b84-11ea-917e-d81c5da836c9.png)
https://pytorch.org/docs/stable/nn.html#logsoftmax

**Comment by bicycleman15** [comment]
Can I take this ?

**Comment by LaRiffle** [comment]
Sure go ahead :)

**Comment by bicycleman15** [comment]
hey @LaRiffle 
I was thinking something along these lines , I might be wrong
```python
def softmax(self , dim ):
        """calculate the softmax along a given dimension
        """
        max_val = self.max(dim, keepdim=True)[0]  # pick the first 
        
        logit = self - max_val
        
        num = logit.exp()
 
        inv = num.sum(dim, keepdim=True)
        
        # generate one
        one = inv * 0 + 1
        
        result = one / inv

        return num * result
```
Could you help me , I just want to make sure I'm going in the right direction.

**Comment by LaRiffle** [comment]
Hey, I've seen your PR! That's a good start, add tests and checkout for possible existing numeric approximations they can help you have great speed-up (like what was done on tanh) :)

**Comment by LaRiffle** [comment]
Hey @bicycleman15 !
I believe your PR is still open and needs tests if you'd like to add them :)

**Comment by bicycleman15** [comment]
Hey @LaRiffle really sorry for the late reply. I haven't had the time to work on this lately, if someone else would like to take this up, then please go ahead.
I am sorry for holding onto this for so long.
